### ISSUE TITLE
Figure out how to access link editing tools fast and easy

### ISSUE THREAD

**Issue by oleq** [ckeditor/ckeditor5-link#17]
A followup of https://github.com/ckeditor/ckeditor5-link/issues/2#issuecomment-240374891.

> How to quickly edit and existing link using only the keyboard? This is important for a11y and for simple UX. E.g. I move selection to the link by using arrow keys. IMO, that shouldn't result in opening the tooltip yet, because that could be irritating. You could, however, open the tooltip (and focus ofc) by pressing CTRL+K. Then Esc must move you back to the editable. However, clicking on the link to place selection inside it should result in opening the tooltip without any additional action.

Also related to https://github.com/ckeditor/ckeditor5-link/issues/10.

---

Two ideas as to how make the user's life easier emerged in the course of previous discussions:
## Open editing panel on click

To speed up link editing, the link panel could show up when the link is clicked (selection is anchored in the link). The panel would **not** appear when selection is changed using the keyboard to prevent the panel from following the caret as users type – this would be a way too much distraction IMO.

![image](https://cloud.githubusercontent.com/assets/1099479/17966136/67d25cc4-6ac0-11e6-9cca-32480b93c932.png)

This behavior is easy to implement and maintain. It also reduces the numbers of clicks needed to start editing to 1. 
## Show contextual toolbar

When the selection has changed (either by clicking or using the keyboard), a contextual panel/toolbar could appear giving an access to the most important features related to the selection context.

For selection anchored in a link, it would be:
- "Link" button, giving an access to the link editing panel.
- "Unlink" button, to simplify https://github.com/ckeditor/ckeditor5-link/issues/10.

The panel would also bring useful shortcuts for other features like image (edit, rotate, flip), list (change type, indent, outdent), table (insert cell, split cell...) etc., depending on the context of the selection.

![image](https://cloud.githubusercontent.com/assets/1099479/17966585/9f8203b6-6ac2-11e6-8ae4-a9cf8600e35f.png)

Pros
- Smart and useful, could actually help users.

Cons
- Yet another distraction following the selection as you type, click etc. It's raining panels.
- Another feature to maintain.

**Comment by fredck** [comment]
I'm for a third option: "open editing panel on selection change":

 * It avoids introducing yet another UI element, the contextual toolbar.
 * The "unlink" option should be available in the panel for existing links (this should be done in any case).
 * Brings the same workflow, no matter if one uses keyboard or mouse.

The only thing is that we need to figure out how to move the focus to the panel when using keyboard only. Maybe `Tab` is a good option.

**Comment by oleq** [comment]
> I'm for a third option: "open editing panel on selection change":

I'm against it, as I mentioned:

>The panel would not appear when selection is changed using the keyboard to prevent the panel from following the caret as users type – this would be a way too much distraction IMO.

The link editing panel would always be visible and sticky to the selection as the user types and navigates in editable. It would stick like a glue, and follow the caret like a dog on a leash, interfere with the **lightweight editing experience that we set up as a goal in v5**. 

It would be especially annoying if there's a selection in a link and the user wants to select (and copy) some content using mouse, which is covered by the panel, even though the user does not want to do anything with links at that particular moment.

> The only thing is that we need to figure out how to move the focus to the panel when using keyboard only. Maybe Tab is a good option.

It isn't. When one writes a text in editable, <kbd>Tab</kbd> key should be reserved for switching to the next focusable component in the web page and, optionally, manipulation in lists (indent/outdent). 

Otherwise, there would be no way to indent list with keyboard in such case

```
* ...
* <p>foo <a>b^ar</a></p> // Tab will focus input in the link editing panel.
* ...
```

and there would be no way to escape the editor in such case (or even get rid of the panel): 

```
<editable><p><a>The onl^y content of editable</a></p></editable>
```

It feels like an accessibility issue for me. 

BTW: We need to rethink <kbd>Tab</kbd> key behaviour in the context of lists too, because it could also get the user stuck in editor if there's no other content in editable but list items.

**Comment by fredck** [comment]
What about mimicking GDocs behavior at this point, with the additional value of being able to click to open the URL in a new tab (which I remember using extensively).

**Comment by Reinmar** [comment]
@fredck: +1. I've proposed this in another thicket as well. The link implementation in GDocs is really nice – can't find anything irritating there.

**Comment by oleq** [comment]
@fredck +1. I'm cool with it. Anything useful can land in that panel in fact, depending on the feature.
![image](https://cloud.githubusercontent.com/assets/1099479/18000730/b428dbe2-6b7f-11e6-8949-0879ff1d3f9a.png)

And the design could be more compact.

![image](https://cloud.githubusercontent.com/assets/1099479/18000741/c22fe92e-6b7f-11e6-8788-65012fe5188f.png)

So it's up to us. I hope this contextual toolbar panel will spread across the features and let us keep the main toolbar more lightweight.

**Comment by oleq** [comment]
BTW:

> The only thing is that we need to figure out how to move the focus to the panel when using keyboard only. Maybe Tab is a good option.

http://ckeditor.github.io/editor-recommendations/features/numbered-list.html says

> Tab key is generally used to steer focus from and into the editor and overwriting its behavior could be misleading. Therefore the following key bindings MAY be used instead of the Tab key[3]:
* Ctrl+M for indenting,
* Ctrl+Shift+M for outdenting.

**Comment by Reinmar** [comment]
We've got two options:

1. If we assign a keystroke to each option in that panel, then you won't need to focus it to access them. This is the simplest option, but means that we have to have keystrokes defined for everything (not bad) and that someone will need to remember them (perhaps not that bad either – as a last resort one can use a mouse or check keystrokes popup which we'll create one day).
2. We can find a keystroke which will allow focusing that toolbar (and every toolbar that we'll have in fact – so there may be a keystroke to cycle through toolbars). This is compatible with CKE5, will be easier to remember for some users, BUT, it doesn't mean that we don't have to assign keystroke to every item anyway. If a toolbar item has a keystroke, then you can use it with one key press. If you need to cycle through toolbars, then cycle through their items... it's not very useful.

So, tl;dr is that a keystroke assigned to each item is a must anyway, so this can be our approach for now. We can later work on accessing toolbars using keystrokes, but this will be more a11y oriented task. General UX IMO will be sorted out by assigning keystrokes and making them discoverable (tooltips).

**Comment by fredck** [comment]
We're probably discussing this in the wrong place, but I'll leave my comment about that `Tab` key here...

 * Yes, we are aware that `Tab` is supposed to cycle focus on elements around the page.
 * Yes, `Tab` MUST be the key used to indent lists. Users may get trapped... maybe move to next element if there is no more room for indentation... or even a secondary key to leave the editor. The fact is that an enormously larger group of users would benefit from having `Tab` as the indentation key, which is btw a de facto standard for it.
 * If we want to have a key to access toolbars, we could use this instead of `Tab` of course, much probably cycling among existing toolbars. For example, we use `Alt+F10` in CKEditor 4. In the CKEditor 5 case, we may use the same key combination... first press will focus the contextual mini-toolbar... second press goes to the main toolbar.
 * `Tab` could instead be used to access the edit panel for easy discoverability, especially if it is placed after the link. There should be no users trapped here because it is enough to `Esc` to close the panel and then `Tab` gets back to its default behavior.

**Comment by Reinmar** [comment]
So would you agree that for now it's enough if we have keystrokes for each item in the link mini-toolbar?

**Comment by fredck** [comment]
> So would you agree that for now it's enough if we have keystrokes for each item in the link mini-toolbar?

I think this is good in any case, but for sure this doesn't solve the accessibility requirement because it doesn't promote the discoverability of features.

We didn't work on keyboard support in any element of the UI so far, so I don't see the reason to work on this right now just for the link feature. We're much probably talking about a editor-wide task for this in the upcoming future. In fact, waiting a bit for it may be beneficial because we'll have all UI elements figured out by then.

**Comment by oskarwrobel** [comment]
Looks like it's done.

![mar-24-2017 16-04-56](https://cloud.githubusercontent.com/assets/1057056/24300132/b9be8792-10ab-11e7-9234-0cbd7fb93724.gif)

**Comment by oleq** [comment]
@oskarwrobel Awaiting https://github.com/ckeditor/ckeditor5-link/issues/31 to be closed.

**Comment by Reinmar** [comment]
I think that the current work on the new theme resolves this ticket.
